### PR TITLE
NDArrays Support Slicing

### PR DESCRIPTION
**Pull request by johnc1231**
This adds slicing support to NDArrays. I've implemented most of what numpy offers here under "Basic Indexing": https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html

Edge cases to note, assuming I have a 3 dimensional ndarray `arr`:

- `arr[:, ::, ::1]` returns the whole array (: is shorthand for "give me the full extent of this particular dimension", 2 colons does the same thing since it's like making a slice with all default values, and 2 colons followed by a 1 is specifying a slice with step size 1, which is the default step size. ). 
- `arr[2:3, 4:8:1, 5:10]` passing in 1 as a step argument has no effect, as 1 is assumed to be the default step. 
- `arr[:, :, -5:-2]` means that in the last dimension we should start selecting at the 5th to last element and stop at the second to last element
- `arr[:, :, 3]` will return a 2 dimensional ndarray, since by passing in a particular value for one of the dimensions I've "deleted" a dimension.
- `arr[:, :, ::-1]` will reverse the elements of last dimension.

I **don't** support:
- Passing in an ellipsis
- Something like `np.newaxis`

^ We can definitely support both of those things, but I haven't done it yet / doesn't seem super high priority

### CODE REVIEW

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 1]
```diff
@@ -3520,9 +3520,9 @@ def __getitem__(self, item):
             slices = []
             for i, s in enumerate(item):
                 if isinstance(s, slice):
-                    start = s.start if s.start is not None else to_expr(0, tint64)
-                    stop = s.stop if s.stop is not None else self.shape[i]
                     step = s.step if s.step is not None else to_expr(1, tint64)
+                    start = hl.cond(s.start >= 0, s.start, self.shape[i] + s.start) if s.start is not None else hl.cond(step >= 0, to_expr(0, tint64), self.shape[i] - 1)
+                    stop = hl.cond(s.stop >= 0, s.stop, self.shape[i] + s.stop) if s.stop is not None else hl.cond(step >= 0, self.shape[i], to_expr(-1, tint64))
```
Comment (johnc1231) [inline hail/python/hail/expr/expressions/typed_expressions.py]: Admittedly these two lines are kind of a lot, but didn't find a clearer way to write them.

**Comment by danking** [comment]
Let's make issues for new axis and ellipsis and tag them ndarray or something so we can keep track of outstanding issues.

**Comment by johnc1231** [comment]
Sure, will do

File: hail/python/test/hail/expr/test_ndarrays.py [hunk 2]
```diff
@@ -61,28 +61,34 @@ def test_ndarray_ref():
...
+        (mat[0, 1:4:2] + mat[:, 1:4:2], np_mat[0, 1:4:2] + np_mat[:, 1:4:2]),
+        (rect_prism[0, 0, -3:-1], np_rect_prism[0, 0, -3:-1]),
+        (flat[15:5:-1], np_flat[15:5:-1]),
+        (flat[::-1], np_flat[::-1]),
+        (flat[::22], np_flat[::22]),
+        (flat[::-22], np_flat[::-22]),
+        (flat[15:5], np_flat[15:5]),
+        (flat[3:12:-1], np_flat[3:12:-1]),
+        (flat[12:3:1], np_flat[12:3:1]),
+        (mat[::-1, :], np_mat[::-1, :])
```
Comment (danking) [inline hail/python/test/hail/expr/test_ndarrays.py]: let's add a test with a negative step that's not a multiple of the slice's size, e.g. `[4:1:-2]`

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 3]
```diff
@@ -2423,6 +2425,61 @@ private class Emit(
...
+              val oneSlice = new CodePTuple(p, region, slicesTuple(dim))
+              codeSlices += ((oneSlice[Long](0), oneSlice[Long](1), oneSlice[Long](2)))
+            }
+            case _: PInt64 => {
+              codeRefMap += dim -> slicesTuple(dim)
+            }
+          }
+        }
+
+        val outputShape = codeSlices.toArray.map{ case (start, stop, step) =>
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: spaces before a curly

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 4]
```diff
@@ -2423,6 +2425,61 @@ private class Emit(
...
+        val slicesTuple = new CodePTuple(coerce[PTuple](slicesIR.pType), region, slicesValueAddress)
+
+        val codeSlices = ArrayBuffer[(Code[Long], Code[Long], Code[Long])]()
+        val codeRefMap = mutable.Map[Int, Code[Long]]()
+
+        coerce[PTuple](slicesIR.pType).types.zipWithIndex.foreach { case (sliceOrIndex, dim) =>
+          sliceOrIndex match {
+            case p: PTuple => {
+              val oneSlice = new CodePTuple(p, region, slicesTuple(dim))
+              codeSlices += ((oneSlice[Long](0), oneSlice[Long](1), oneSlice[Long](2)))
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: what about missing tuple elements?

File: hail/python/hail/expr/expressions/typed_expressions.py [hunk 5]
```diff
@@ -3520,9 +3520,9 @@ def __getitem__(self, item):
             slices = []
             for i, s in enumerate(item):
                 if isinstance(s, slice):
-                    start = s.start if s.start is not None else to_expr(0, tint64)
-                    stop = s.stop if s.stop is not None else self.shape[i]
                     step = s.step if s.step is not None else to_expr(1, tint64)
```
Comment (danking) [inline hail/python/hail/expr/expressions/typed_expressions.py]: we should produce a useful error when `step` is zero, yeah?

File: hail/python/test/hail/expr/test_ndarrays.py (hunk 2 above)
Comment (danking) [reply hail/python/test/hail/expr/test_ndarrays.py]: and `[0:0:1]` which should return the empty array.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 6]
```diff
@@ -2423,6 +2425,61 @@ private class Emit(
...
+            case _: PInt64 => {
+              codeRefMap += dim -> slicesTuple(dim)
+            }
+          }
+        }
+
+        val outputShape = codeSlices.toArray.map{ case (start, stop, step) =>
+          (step >= 0L && start <= stop).mux(
+            const(1L) + ((stop - start) - const(1L)) / step,
+            (step < 0L && stop <= start).mux((((stop - start) + const(1L)) / step) + const(1L),
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: I think this is easier to read with the consequent on the next line.

Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: `[-4:-1:2]`
=>
```
((-1 + 4) + 1)/2 + 1
= 4/2 + 1
= 3
```
seems wrong, right?

Comment (danking) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: We should get just -4 and -2.
```
In [20]: [1,2,3,4,5][-4:-1:2]                                                                                                                                                
Out[20]: [2, 4]
```

Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: In the example you give, `[1, 2, 3, 4, 5]`, the length is 5, so in python I convert the slice `[-4:-1:2]` to `[1:4:2]`, which will work. That's line 3524 of typed_expressions.py

Comment (danking) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: So how do we get `stop <= start` in Scala?

Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Something like `[5:1:-1]`

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 7]
```diff
@@ -2430,6 +2432,71 @@ private class Emit(
...
+
+        coerce[PTuple](slicesIR.pType).types.zipWithIndex.foreach { case (sliceOrIndex, dim) =>
+          sliceOrIndex match {
+            case p: PTuple => {
+              val oneSlice = new CodePTuple(p, region, slicesTuple(dim))
+              missingSliceElements = missingSliceElements || oneSlice.isMissing(0) || oneSlice.isMissing(1) || oneSlice.isMissing(2)
+              codeSlices += ((oneSlice[Long](0), oneSlice[Long](1), oneSlice[Long](2)))
+            }
+            case _: PInt64 => {
+              codeRefMap += dim -> slicesTuple(dim)
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: this is `codeRefMap(dim) = slicesTuple(dim)`

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 8]
```diff
@@ -2430,6 +2432,71 @@ private class Emit(
...
+        val codeSlices = ArrayBuffer[(Code[Long], Code[Long], Code[Long])]()
+        val codeRefMap = mutable.Map[Int, Code[Long]]()
+
+        var missingSliceElements = const(false)
+
+        coerce[PTuple](slicesIR.pType).types.zipWithIndex.foreach { case (sliceOrIndex, dim) =>
+          sliceOrIndex match {
+            case p: PTuple => {
+              val oneSlice = new CodePTuple(p, region, slicesTuple(dim))
+              missingSliceElements = missingSliceElements || oneSlice.isMissing(0) || oneSlice.isMissing(1) || oneSlice.isMissing(2)
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: this is `missingSliceElements ||= ...`

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 9]
```diff
@@ -2430,6 +2432,71 @@ private class Emit(
...
+        val slicest = emit(slicesIR, env, resultRegion, None)
+        val slicesValueAddress = mb.newField[Long]
+        val slicesTuple = new CodePTuple(coerce[PTuple](slicesIR.pType), region, slicesValueAddress)
+
+        val codeSlices = ArrayBuffer[(Code[Long], Code[Long], Code[Long])]()
+        val codeRefMap = mutable.Map[Int, Code[Long]]()
+
+        var missingSliceElements = const(false)
+
+        coerce[PTuple](slicesIR.pType).types.zipWithIndex.foreach { case (sliceOrIndex, dim) =>
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: I think this will read more clearly with some helper functions:
```scala
class CodePTuple(...) {
  ...
  def withTypes = withTypesAndIndex.map(x => x._1 -> x._2)

  def withTypesAndIndex = (0 until pType.nFields).map(i => (pType.types(i), apply(i), i))

  def missingness = (0 until pType.nFields).map(isMissing _)

  // we can write this out for sizes 1 through 3
  def values[T, U, V] = {
    assert pType.nFields == 3
    (apply[T](0), apply[U](1), apply[V](2))
  }
}
```
and:
```scala
slicesTuple.withTypesAndIndex {
  case (p: PTuple, _slice, _) =>
    val slice = new CodePTuple(p, region, _slice)
    missingElements ||= slice.missingness.fold(_ || _)
    codeSlices += slice.values[Long, Long, Long]
  case (_: PInt64, index, dim) =>
    codRefMap(dim) = index
```

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 7 above)
Comment (danking) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Also need to check the missingness of this int.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 10]
```diff
@@ -2430,6 +2432,71 @@ private class Emit(
           }
         }
 
+      case x@NDArraySlice(child, slicesIR) =>
+        val childEmitter = deforest(child)
+
+        val slicest = emit(slicesIR, env, resultRegion, None)
+        val slicesValueAddress = mb.newField[Long]
+        val slicesTuple = new CodePTuple(coerce[PTuple](slicesIR.pType), region, slicesValueAddress)
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: I think we should just call this `slices`, we know it is a tuple from its type.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 8 above)
Comment (danking) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: This also misses if the tuple element itself is missing.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 9 above)
Comment (danking) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Actually might be more clear as two separate steps:
```scala
val slicers = slicesTuple.withTypes.collect {
    case (t: PTuple, slice) => new CodePTuple(t, region, slice) }
val codeSlices = slicers.map(_.values[Long, Long, Long])

val indexers = slicesTuple.withTypesAndIndex.collect {
    case (_: PInt64, indexer, i) => i => coerce[Long](indexer) }
val codeRefMap = indexers.toMap

val anyMissingness = slicesTuple.missingness.fold(_ || _) ||
    slicers.map(_.missingness.fold(_ || _)).fold(_ || _)
```

Comment (danking) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: I prefer a more descriptive name for `codeRefMap`, maybe something like `indexer`? `selector`? `nonSlicer`?

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 11]
```diff
@@ -2430,6 +2432,71 @@ private class Emit(
...
+              codeRefMap += dim -> slicesTuple(dim)
+            }
+          }
+        }
+
+        val outputShape = codeSlices.toArray.map { case (start, stop, step) =>
+          (step >= 0L && start <= stop).mux(
+            const(1L) + ((stop - start) - const(1L)) / step,
+            (step < 0L && stop <= start).mux(
+              (((stop - start) + const(1L)) / step) + const(1L),
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: We should keep the order consistent between these two branches for all of our sanity.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 12]
```diff
@@ -2430,6 +2432,71 @@ private class Emit(
           }
         }
 
+      case x@NDArraySlice(child, slicesIR) =>
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: We need a check somewhere that `slicesIR.pType.nFields == child.pType.nDims`. I don't know what the current status of InferType vs InferPType is, but that would be the place, IMO.

Comment (danking) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: I don't see such a check currently.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 11 above)
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: As long as numbers are the RHS, the const shouldn't be necessary.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 13]
```diff
@@ -2430,6 +2432,71 @@ private class Emit(
...
+              missingSliceElements = missingSliceElements || oneSlice.isMissing(0) || oneSlice.isMissing(1) || oneSlice.isMissing(2)
+              codeSlices += ((oneSlice[Long](0), oneSlice[Long](1), oneSlice[Long](2)))
+            }
+            case _: PInt64 => {
+              codeRefMap += dim -> slicesTuple(dim)
+            }
+          }
+        }
+
+        val outputShape = codeSlices.toArray.map { case (start, stop, step) =>
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Any reason to have the toArray before the map? I think it reads more clearly to coerce to the type of interest as the last step.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 14]
```diff
@@ -2430,6 +2432,71 @@ private class Emit(
...
+
+        val setupMissing = Code(
+          slicest.setup,
+          slicesValueAddress := slicest.value[Long],
+          childEmitter.setupMissing
+        )
+
+        val setupShape = Code(
+          childEmitter.setupShape
+        )
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: remove the  unnecessary `Code`

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 15]
```diff
@@ -2430,6 +2432,71 @@ private class Emit(
...
+
+            val sliceIdxVars = Array.tabulate(childEmitter.nDims) { dim =>
+              if (codeRefMap.contains(dim)) {
+                codeRefMap(dim)
+              } else {
+                val (start, _, step) = sliceIdxVarsIter.next()
+                val oldIdxVar = oldIdxVarsIter.next()
+                start + oldIdxVar * step
+              }
+            }
```
Comment (danking) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: I think this bit would be more clear if transformed the wider bit to look more
like this:
```scala
val slicers = slicesTuple.withTypes.collect {
    case (t: PTuple, slice) => new CodePTuple(t, region, slice) }
val anyMissingness = slicesTuple.missingness.fold(_ || _) ||
    slicers.map(_.missingness.fold(_ || _)).fold(_ || _)
val shape = slicers.map(_.value[Long, Long, Long]).map { case (start, stop, step) =>
  ...
}.toArray

// ...

new ...
  override def outputElement ....
    val inIdx = idxVars.iterator
    val sliceIdxVars = slicesTuple.withTypes.map {
      case (_: PInt64, indexer, i) =>
        coerce[Long](indexer)
      case (t: PTuple, slice) =>
        val (start, _, step) = val CodePTuple(t, region, slice).value[Long, Long, Long]
        start + inIdx.next * step
    }
    childEmitter.outputElement(sliceIdxVars)
```
I think this is conceptually a bit easier to follow

File: hail/python/hail/expr/expressions/typed_expressions.py (hunk 1 above)
Comment (danking) [reply hail/python/hail/expr/expressions/typed_expressions.py]: They might be clearer if not written with the ternary operator and instead as
```python3
if ...:
  step = ..
else:
  step = ...
```
But I don't have strong feelings here.

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala (hunk 12 above)
Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Added it in `InferType`

**Comment by johnc1231** [comment]
All addressed
